### ISSUE TITLE
Can't delete file from SD card

### ISSUE THREAD

**Issue by hugleo** [koreader/koreader#4872]
Device: Likebook Mars
Koreader version: last nighty build

Koreader can't delete files from a external storage sd card. Tried another third party app and is deleting normally. 
From internal storage Koreader is deleting normally.

![Screenshot_20190403-191050](https://user-images.githubusercontent.com/4573505/55516721-74de4780-5644-11e9-9c04-22d052816a95.png)

**Comment by pazos** [comment]
The same happens on the emulator. This affects android M and higher (6.0+)

```logcat
04-04 00:30:13.283 16725 16741 W KOReader: command /system/bin/mkdir /storage/19E8-0A11/test returned 1
04-04 00:30:13.283 16725 16741 V KOReader:  stdout:
04-04 00:30:13.283 16725 16741 V KOReader:  stderr: mkdir: '/storage/19E8-0A11/test': Permission denied
```
@hugleo: feel free to work on this if you wish. I have no special interest on this issue. Some background: https://commonsware.com/blog/2017/11/15/storage-situation-removable-storage.html

**Comment by hugleo** [comment]
I found out that as kobo use lfs mkdir and android is not very posix friendly and the way to do it that android force the user select a directory tree using Java API. I don't think it's worth since less Java exposed to the Koreader backend is better. Even so very few e-readers devices support a external sd card expansion and many of them already are providing a 32 GB+ memory. Another way is to root the device and grant the permission in the /system/etc/permissions/platform.xml file.
But for now I use a workaround and move my book files to /SanDisk SD card/Android/data/org.koreader.launcher/files. Even if that is a external sd card koreader is able to write in this folder. I just need to remember that If I uninstall koreader I need to eject the sd card first otherwise the /SanDisk SD card/Android/data/org.koreader.launcher folder will be removed ;-)

**Comment by pazos** [comment]
> I found out that as kobo use lfs mkdir and android is not very posix friendly

True, android is not ~~posix~~ friendly.

> The way to do it that android force the user select a directory tree using Java API. I don't think it's worth since less Java exposed to the Koreader backend is better

I don't know if you are aware of https://github.com/koreader/android-luajit-launcher, where the interface between java and lua happens. I think that I understand what you said but we are currently using a bunch of things only available in java, so no problem here.

>  Very few e-readers devices support a external sd card expansion and many of them already are providing a 32 GB+ memory.

True again, but a lot of phones/tablets have sdcard support.

> Another way is to root the device and grant the permission in the /system/etc/permissions/platform.xml file.

I think that is great info for the wiki but not helpful for most users and should not be a requirement for a specific function.

> But for now I use a workaround and move my book files to /SanDisk SD card/Android/data/org.koreader.launcher/files. Even if that is a external sd card koreader is able to write in this folder. I just need to remember that If I uninstall koreader I need to eject the sd card first otherwise the /SanDisk SD card/Android/data/org.koreader.launcher folder will be removed ;-)

You can store the books in any folder inside /sdcard and they won't be removed if you uninstall the program. /sdcard has read/write permissions on all supported android versions (4-9).

**Comment by hugleo** [comment]
> > I found out that as kobo use lfs mkdir and android is not very posix friendly
> 
> True, android is not posix friendly.
> 
> > The way to do it that android force the user select a directory tree using Java API. I don't think it's worth since less Java exposed to the Koreader backend is better
> 
> I don't know if you are aware of https://github.com/koreader/android-luajit-launcher, where the interface between java and lua happens. I think that I understand what you said but we are currently using a bunch of things only available in java, so no problem here.
> 
Yes, I know about that. All writes calls lines like lfs.mkdir(sidecar) would have a android variant. But now you talk looks less harmful that before ;-) 

> > Very few e-readers devices support a external sd card expansion and many of them already are providing a 32 GB+ memory.
> 
> True again, but a lot of phones/tablets have sdcard support.
> 
> > Another way is to root the device and grant the permission in the /system/etc/permissions/platform.xml file.
> 
> I think that is great info for the wiki but not helpful for most users and should not be a requirement for a specific function.
> 
> > But for now I use a workaround and move my book files to /SanDisk SD card/Android/data/org.koreader.launcher/files. Even if that is a external sd card koreader is able to write in this folder. I just need to remember that If I uninstall koreader I need to eject the sd card first otherwise the /SanDisk SD card/Android/data/org.koreader.launcher folder will be removed ;-)
> 
> You can store the books in any folder inside /sdcard and they won't be removed if you uninstall the program. /sdcard has read/write permissions on all supported android versions (4-9).

It's true if you use a SD card formatted as internal memory. This way Koreader will write with no problems. But if the card is formatted to use as portable then Koreader will be able to read all files however just the folder /storage/XXXX/data/org.koreader.launcher will have the write permission. Android system automatically creates that folder with the application name in sd card too , not just in internal memory. A koreader reinstall looks like is keeping the folder but this folder will be deleted if koreader is uninstalled, that's is because I eject the card if I uninstalled Koreader (sometimes need If I use my private key to the apk) so android system I'll not execute the delete action :)

**Comment by hugleo** [comment]
More background, add for future use: "since Android 4.4 (API level 19) introduces the Storage Access Framework (SAF)"
https://developer.android.com/guide/topics/providers/document-provider

**Comment by pazos** [comment]
> It's true if you use a SD card formatted as internal memory. This way Koreader will write with no problems. But if the card is formatted to use as portable then Koreader will be able to read all files however just the folder /storage/XXXX/data/org.koreader.launcher will have the write permission. Android system automatically creates that folder with the application name in sd card too , not just in internal memory

[android.externalStorage](https://github.com/koreader/android-luajit-launcher/blob/master/assets/android.lua#L1412-L1427) should return the primary storage path. It doesn't matter if is an internal partition or an external sdcard formatted as internal storage. This path is something like `/storage/emulated/0` on modern versions, and it should be symlinked to `/sdcard`. It is available in the environment as EXTERNAL_STORAGE. That's the path where we got RW permissions granted and everything there should stay as is even if you remove the application(s). That does not applies to anything stored in $EXTERNAL_STORAGE/Android/*

Secondary storage paths, if they're available, should be present on the OS environment too. You can retrieve it with `adb shell printenv`.

**Comment by pazos** [comment]
Closing this issue, since it cannot be fixed without root permissions on Android 4.3+ - Android 5.0 and cannot be fixed at all on Android 6.0+. 

A workaround is, like @hugleo did, format external sdcard as internal storage. As long as we store books inside /sdcard (which should be a link to /storage/emulated/0 or the like) we can grant read&write permissions.

For using books from other sources, we should implement the Storage Access Framework, but that doesn't change the fact that **cannot delete files outside /sdcard"
